sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PropertyNamesShouldNotMatchGetMethods.cs

```diff
@@ -0,0 +1,113 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class PropertyNamesShouldNotMatchGetMethods : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4059";
+        private const string MessageFormat = "Change either the name of the property '{0}' or the name of " +
+            "the method '{1}' to make them distinguishable.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var classSymbol = c.SemanticModel.GetDeclaredSymbol(c.Node) as INamedTypeSymbol;
+                    if (classSymbol == null)
+                    {
+                        return;
+                    }
+
+                    var classMembers = classSymbol.GetMembers().Where(IsPublicOrProtectedMember);
+                    var properties = classMembers.OfType<IPropertySymbol>().Where(property => !property.IsOverride);
+                    var methods = classMembers.OfType<IMethodSymbol>().ToList();
+
+                    foreach (var collidingMembers in GetCollidingMembers(properties, methods))
+                    {
+                        var propertyIdentifier = collidingMembers.Item1;
+                        var methodIdentifier = collidingMembers.Item2;
+
+                        c.ReportDiagnostic(Diagnostic.Create(
+                            rule,
+                            propertyIdentifier.GetLocation(),
+                            additionalLocations: new[] { methodIdentifier.GetLocation() },
+                            messageArgs: new[] { propertyIdentifier.ValueText, methodIdentifier.ValueText }));
+                    }
+                }, SyntaxKind.ClassDeclaration);
+        }
+
+        private static bool IsPublicOrProtectedMember(ISymbol symbol)
```

Comment:
Could drop "Member" - e.g. rename to "IsPublicOrProtected". This code is not specific to member symbol.